### PR TITLE
meta.yaml keys for cross-build environment

### PR DESCRIPTION
**Pull request by hoodmane**
This moves unisolation into a package key `cross-build-env: true` means the package is part of the cross build environment and should be unisolated. `cross-build-files` gives a list of files that should be copied from the built copy of the package into the host copy of the package.

This will allow us to construct a cross build environment automatically as part of building packages. If we have these files and the Python `include` directory, this is sufficient for cross-building packages binary packages.

### CODE REVIEW

File: pyodide-build/pyodide_build/buildpkg.py [hunk 1]
```diff
@@ -145,6 +146,7 @@ def get_bash_runner():
             "SIDE_MODULE_LDFLAGS",
             "STDLIB_MODULE_CFLAGS",
             "OPEN_SSL_ROOT",
+            "UNISOLATED_PACKGES",
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/buildpkg.py:149]:
```suggestion
            "UNISOLATED_PACKAGES",
```

File: pyodide-build/pyodide_build/common.py [hunk 2]
```diff
@@ -203,3 +202,46 @@ def search_pyodide_root(curdir: str | Path, *, max_depth: int = 5) -> Path:
...
+def get_pyodide_root():
+    init_environment()
+    return Path(os.environ["PYODIDE_ROOT"])
+
+
+@functools.cache
+def get_unisolated_packages():
+    import json
+
+    if "UNISOLATED_PACKGES" in os.environ:
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/common.py:238]:
```suggestion
    if "UNISOLATED_PACKAGES" in os.environ:
```

File: pyodide-build/pyodide_build/common.py [hunk 3]
```diff
@@ -203,3 +202,46 @@ def search_pyodide_root(curdir: str | Path, *, max_depth: int = 5) -> Path:
...
+        return json.loads(os.environ["UNISOLATED_PACKGES"])
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/common.py:239]:
```suggestion
        return json.loads(os.environ["UNISOLATED_PACKAGES"])
```

File: pyodide-build/pyodide_build/common.py [hunk 4]
```diff
@@ -203,3 +202,46 @@ def search_pyodide_root(curdir: str | Path, *, max_depth: int = 5) -> Path:
...
+    PYODIDE_ROOT = get_pyodide_root()
+    unisolated_packages = []
+    for pkg in (PYODIDE_ROOT / "packages").glob("**/*.meta"):
+        config = parse_package_config(pkg, False)
+        if config.get("build", {}).get("cross-build-env", False):
+            unisolated_packages.append(config["package"]["name"])
+    os.environ["UNISOLATED_PACKGES"] = json.dumps(unisolated_packages)
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/common.py:246]:
```suggestion
    os.environ["UNISOLATED_PACKAGES"] = json.dumps(unisolated_packages)
```

File: pyodide-build/pyodide_build/common.py [hunk 5]
```diff
@@ -203,3 +202,46 @@ def search_pyodide_root(curdir: str | Path, *, max_depth: int = 5) -> Path:
...
+
+@functools.cache
+def get_unisolated_packages():
+    import json
+
+    if "UNISOLATED_PACKGES" in os.environ:
+        return json.loads(os.environ["UNISOLATED_PACKGES"])
+    PYODIDE_ROOT = get_pyodide_root()
+    unisolated_packages = []
+    for pkg in (PYODIDE_ROOT / "packages").glob("**/*.meta"):
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/common.py:242]: Shouldn't this be:

```suggestion
    for pkg in (PYODIDE_ROOT / "packages").glob("**/meta.yaml"):
```

It's interesting why it works...?

File: pyodide-build/pyodide_build/pypabuild.py [hunk 6]
```diff
@@ -62,7 +60,7 @@ def install_reqs(env: IsolatedEnv, reqs: set[str]):
     # only recythonize if it is present. We need them to always recythonize so
     # we always install cython. If the reqs included some cython version already
     # then this won't do anything.
-    env.install(["cython"])
+    env.install(["cython", "pythran"])
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/pypabuild.py:63]: If pythran is installed here, do we still need to install it in scipy?

**Review by ryanking13 — COMMENTED** [review]
Thanks, @hoodmane! I have some minor comments.

File: pyodide-build/pyodide_build/pypabuild.py (hunk 6 above)
Comment (hoodmane) [reply pyodide-build/pyodide_build/pypabuild.py:63]: This change should be reverted for now because it doesn't actually work.

**Comment by hoodmane** [comment]
Thanks for the review @ryanking13!

File: pyodide-build/pyodide_build/pypabuild.py (hunk 6 above)
Comment (hoodmane) [reply pyodide-build/pyodide_build/pypabuild.py:63]: Currently the issue is that we are using the `host_site_packages` one for the C include path but importing from the pypa/build virtual environment. The problem is that we have no obvious way of pointing the C include path into the pypa/build virtual environment.

**Review by ryanking13 — APPROVED** [review]
Thanks, @hoodmane!

**Comment by hoodmane** [comment]
Thanks for the review @ryanking13! We are very close to getting out of tree builds to work.

**Pull request merged**